WL-4274 Modified calendar entity to support calendar component in Lesson

ouit0408 opened the pull request:
Exposed method 'getCalendarReferences' in CalendarService to get
references from all calendars for a site.
Added attributes 'description' and 'eventImage' in the EventSummary class
which will be used in the json returned by CalendarEventEntityProvider.
Created map for event type and image in the CalendarUtil class which can be
accessed using method 'getEventImageMap'.
CalendarEventEntityProvider's 'getCalendarEventsForSite' will now return
events from all the merged calendars for a site.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -112,21 +120,23 @@ public boolean entityExists(final String id) {
-			cal = calendarService.getCalendar(String.format(
-					"/calendar/calendar/%s/main", siteId));
+			//get all events from the merged calendars
+			CalendarEventVector calendarEventVector = calendarService.getEvents(calendarService.getCalendarReferences(siteId), getYearTimeRange(new CalendarUtil()));
```
I wonder if including the merged calendars should be an option on the URL as if we push this change back to Sakai, someone else who is consuming the API will start getting events they might not have been expecting. WDYT?

ouit0408 replied:
I got your point, so I should rather add some additional parameter in the url when called from Lessons and check here if the request is for merged calendars or not?

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -219,4 +229,32 @@ public CalendarEventDetails getCalendarEventDetails(EntityView view) {
+
+		calendarUtil.setDay(calendarUtil.getYear(), calendarUtil.getMonthInteger(), 1);
+		int numberOfCurrentDays = calendarUtil.getNumberOfDays();
+		int tempCurrentMonth = calendarUtil.getMonthInteger();
```
What's the point of this variable?

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -219,4 +229,32 @@ public CalendarEventDetails getCalendarEventDetails(EntityView view) {
+		calendarUtil.setDay(calendarUtil.getYear(), calendarUtil.getMonthInteger(), 1);
+		int numberOfCurrentDays = calendarUtil.getNumberOfDays();
+		int tempCurrentMonth = calendarUtil.getMonthInteger();
+		int tempCurrentYear = calendarUtil.getYear();
```
What's the point of this variable?

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -219,4 +229,32 @@ public CalendarEventDetails getCalendarEventDetails(EntityView view) {
+		int dayofweek = 0;
+
+		calendarUtil.setDay(calendarUtil.getYear(), calendarUtil.getMonthInteger(), 1);
+		int numberOfCurrentDays = calendarUtil.getNumberOfDays();
```
What's the point of this variable?

buckett commented on calendar-util/util/src/java/org/sakaiproject/util/CalendarUtil.java:
```diff
@@ -579,5 +586,27 @@ static String getLocalPMString(DateTime now) {
+		if(eventIconMap.size() == 0){
+			//load keys for locale root
+			Map<String, String> eventKeyMap = new HashMap<String, String>();
+			ResourceBundle bundle = ResourceBundle.getBundle("calendar", Locale.ROOT);
```
Doesn't this mean that the `alt` messages are always in the root locale, rather than the user's locale?

ouit0408 replied:
In 'chef_calendar.vm' alt for the image is read from the user's locale so it displays correct alt message.

buckett commented on calendar-impl/impl/src/java/org/sakaiproject/calendar/impl/BaseCalendarService.java:
```diff
@@ -7413,8 +7413,14 @@ protected void handleAccessOpaqueUrl(HttpServletRequest req, HttpServletResponse
-	protected List<String> getCalendarReferences(String siteId) {
+
+	/**
+	 * Made this method public as the same code was repeated in CalendarBean.java and
```
This shoudn't be in the JavaDoc, but in the commit comment, the JavaDoc should describe what the method does.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -219,4 +229,32 @@ public CalendarEventDetails getCalendarEventDetails(EntityView view) {
 	}
 
+	/**
+	 * Returns TimeRange for an year.
```
This should be in the `@return` and the description should  be of what this is trying to achieve, eg. Builds a timerange that goes from

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -219,4 +229,32 @@ public CalendarEventDetails getCalendarEventDetails(EntityView view) {
+
+		// get the index of the first day in the month
+		int firstDay_of_Month = calendarUtil.getDay_Of_Week(true) - 1;
+		calendarUtil.setPrevDate(firstDay_of_Month);
```
Is this attempting to include the whole week from the start day?

ouit0408 commented:
Implemented review comments and have pushed the changes.

buckett commented on calendar-tool/tool/src/webapp/vm/calendar/chef_calendar.vm:
```diff
@@ -54,49 +54,8 @@ calendar Portlet
-<img src = "#imageLink("$config.getProperty('legend.icon19')")" alt="$tlang.getString('legend.key19')" border="0">
-#elseif ($eventType =="Formative Assessment")
-<img src = "#imageLink("$config.getProperty('legend.icon20')")" alt="$tlang.getString('legend.key20')" border="0">
+<img src = "#imageLink($eventIconMap.get($eventType))" alt="$eventType" border="0">
```
Doesn't this mean the `alt` text is now just the key? eg: legend.icon15

ouit0408 replied:
alt text will have the value passed to the macro 'iconImage($eventType)', why it would have legend.icon? I am confused

buckett replied:
Ok, so eventIconMap has a key of "Activity" and a value of "/icon/activity.jpg", my mistake, all ok here.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -111,23 +121,62 @@ public boolean entityExists(final String id) {
+					calendarUtil = new CalendarUtil(calendar);
+				}
+				//get all events from the merged calendars
+				CalendarEventVector calendarEventVector = calendarService.getEvents(calendarService.getCalendarReferences(siteId), getMonthTimeRange(calendarUtil));
```
This changes the default behaviour of the service again from returning all events in the calendar to returning just a months. Anyone who is using the API already and upgrades would see a change in behaviour.

buckett replied:
Ah, no only if it's merged. But it seems strange that the date limitation only is in effect if you ask for a merged calendar.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -111,23 +121,62 @@ public boolean entityExists(final String id) {
-		} catch (PermissionException e) {
-			throw new EntityNotFoundException("No access to site: " + siteId,
-					siteId);
+				for (Object o : cal.getEvents(null, null)) {
```
I don't think you should be having two loops as this one doesn't have the images set on it.

buckett replied:
Still waiting on this.

ouit0408 replied:
this was the original code present in the class and the trunk has similar code in the new file too.

ouit0408 commented:
updated the PR and have pushed changes.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -94,95 +102,72 @@ public boolean entityExists(final String id) {
 		try {
-			cal = calendarService.getCalendar(String.format(
-					"/calendar/calendar/%s/main", siteId));
+			config.load(getClass().getResourceAsStream("/org/sakaiproject/calendar/tool/calendar.config"));
```
You shouldn't do initialisation on a request like this, while this shouldn't have threading issues it can lead to them down the line, it's much cleaner todo it in an `init()`.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -191,32 +176,141 @@ public boolean entityExists(final String id) {
 		// get siteId
-		String siteId = view.getPathSegment(2);
-		String eventId = view.getPathSegment(3);
+		final String siteId = view.getPathSegment(2);
```
Why all the changes to final?

ouit0408 replied:
new class in sakai trunk has this change so I kept my added code in line with that.

buckett replied:
Cool.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -191,32 +176,141 @@ public boolean entityExists(final String id) {
 		if (StringUtils.isBlank(siteId)) {
 			throw new IllegalArgumentException(
-					"siteId must be set in order to get the news feeds for a site, via the URL /news/site/siteId");
+					"siteId must be set in order to get the calendar feeds for a site, via the URL /calendar/site/siteId");
```
Good catch.

buckett commented on calendar-tool/tool/src/java/org/sakaiproject/calendar/entityproviders/CalendarEventEntityProvider.java:
```diff
@@ -191,32 +176,141 @@ public boolean entityExists(final String id) {
 			throw new EntityNotFoundException("Invalid siteId: " + siteId,
 					siteId);
-		} catch (PermissionException e) {
+		} catch (final PermissionException e) {
```
More final?

ouit0408 committed:
WL-4274 Modified calendar entity to support calendar component in Lessons.

Exposed method 'getCalendarReferences' in CalendarService to get
references from all calendars for a site.

Added attributes 'description' and 'eventImage' in the EventSummary class
which will be used in the json returned by CalendarEventEntityProvider.

Created map for event type and image in the CalendarUtil class which can be
accessed using method 'getEventImageMap'.

CalendarEventEntityProvider's 'getCalendarEventsForSite' will also return
events from all the merged calendars for a site depending on the parameters
in the request.If 'merged' parameter is set method will return events from
all merged calendars else events from the current site will be displayed.
Pulled latest code for class 'CalendarEventEntityProvider' from Sakai trunk

ouit0408 commented:
Added init method to entity provider.
